src/DynamoCore/Graph/Nodes/NodeModel.cs

```diff
@@ -73,6 +73,11 @@ public abstract class NodeModel : ModelBase, IRenderPackageSource<NodeModel>, ID
         /// </summary>
         public virtual string CreationName { get { return this.Name; } }
 
+        /// <summary>
+        /// Returns all the upstream nodes for the given node
+        /// </summary>
+        public HashSet<NodeModel> Upstream = new HashSet<NodeModel>();
```

Comment:
I think we need to clarify in the comment that this property will not return all upstream nodes for a given node unless they have been precomputed for this node using ComputeUpstream and that is the responsibility of the caller.